### ISSUE TITLE
Live-reloading layouts when using ProMotion

### ISSUE THREAD

**Issue by dchersey** [motion-kit/motion-kit#131]
I am using promotion with motion-kit and would LOVE to be able to live-reload changes to my layouts in motion-kit.  

They provide a method for customizing the live-reloading [here] (http://promotion.readthedocs.org/en/master/Guides/Live%20Screen%20Reloading/) and I imagine I could add code to reload the whole layout (what would that be by the way? I don't think reapply! does enough).

The problem is, the layouts don't recompile when promotion live reloading fires. 

I can ask this question over there but I was wondering if you have already given it some thought?

**Comment by jamonholmgren** [comment]
@dchersey You can add your own live reloader, like this:

https://github.com/infinitered/ProMotion/blob/a40c03ca89e21c21ef4c2fc35d5d3b85ea4c77de/lib/ProMotion/repl/repl.rb#L10-L18

I think it would look something like this:

```ruby
        @screen_watcher = LiveReloader.new("app/layouts/**/*.rb", opts).watch do |reloaded_file, new_code, class_names|
          vcs = pm_all_view_controllers(UIApplication.sharedApplication.delegate.window.rootViewController)
          vcs.each do |vc|
            if true # somehow determine if this view controller uses the reloaded layout
              puts "Sending :on_live_reload to #{vc.inspect}." if opts[:debug]
              vc.send(:on_live_reload) if vc.respond_to?(:on_live_reload)
            end
          end
        end
```

So, all you're really doing is watching an additional directory, then doing the same thing that the PM screen watcher does. You'll have to supply your own logic to determine whether the screen uses that layout. It shouldn't be too hard, I'd imagine.

Then, when you've done that, you could remove all of the subviews and re-call `on_load` to re-trigger building your layout.

**Comment by dchersey** [comment]
Sounds great, I will give it a shot!  Thanks @jamonholmgren for all you do :)

Once I have this working, any thoughts on how to cleanly integrate this with promotion besides patching the code? Perhaps a register_watcher api that I could use from an initializer or possibly contribute to motion-kit?

**Comment by colinta** [comment]
Yeah if you can make it generic, let's add it it!  It might be easiest as an external gem, we can include it in this organization.

**Comment by dchersey** [comment]
I have this working -- setting up a PR for Promotion to continue discussion.   It's not going to do anything for projects not using motion-kit so I am not sure there's tremendous value in making it pluggable -- will reference the PR here when it's ready to continue discussion. 

for now the simple addition looks like: 

```ruby

        @layout_watcher = LiveReloader.new("app/layouts/**/*.rb", opts).watch do |reloaded_file, new_code, class_names|
          vcs = pm_all_view_controllers(UIApplication.sharedApplication.delegate.window.rootViewController)
          vcs.each do |vc|
            if pm_is_layout?(vc, class_names) 
              puts "Sending :on_live_reload to #{vc.inspect}." #if opts[:debug]
              vc.send(:on_live_reload) if vc.respond_to?(:on_live_reload)
            end
          end
        end

        "Live reloading of screens, views, and layouts is now on."
```

and

```ruby
    def pm_is_layout?(vc, layout_code)
      definition = layout_code.detect {|e| e =~ /class\s*(\S*)Layout/}
      screen_name = "#{$1}Screen"
      vc.class.to_s == screen_name
    end
```
Note it requires you use naming conventions for the layout -- if I could easily access the source for the vc I could improve that.

**Comment by jamonholmgren** [comment]
Nice, I'll go check it out on ProMotion.

**Comment by jamonholmgren** [comment]
I agree that making it more generic would be best.

**Comment by dchersey** [comment]
Sorry; got caught up; will post PR tomorrow.
